### PR TITLE
KOGITO-3228: add outcome ids to saliency model

### PR DESCRIPTION
**Pull request by kostola**
[JIRA task](https://issues.redhat.com/browse/KOGITO-3228)

### CODE REVIEW

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java [hunk 1]
```diff
@@ -61,17 +62,17 @@ public ExplanationServiceImpl(
                 request.getExecutionId(),
                 request.getModelIdentifier().getResourceType(),
                 request.getModelIdentifier().getResourceId());
-        Prediction prediction = getPrediction(request.getInputs(), request.getOutputs());
-        return localExplainer.explainAsync(prediction, predictionProvider)
+        return CompletableFuture.supplyAsync(() -> getPrediction(request.getInputs(), request.getOutputs()))
+                .thenCompose(prediction -> localExplainer.explainAsync(prediction, predictionProvider))
```
Comment (danielezonca) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java]: What is the reason of this change?
`getPrediction` is plain local method and `localExplainer.explainAsync` already create the `CompletableFuture` internally.
I don't see the reason to add this additional wrapper. Btw you should not use any `*Async` methods without providing the executor in a managed environment

File: explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ExplanationServiceImplTest.java [hunk 2]
```diff
@@ -57,27 +58,49 @@ void init() {
...
+        ExplainabilityResultDto resultDto = assertDoesNotThrow(() ->
+                explanationService.explainAsync(REQUEST, predictionProviderMock)
+                        .toCompletableFuture()
+                        .get(Config.INSTANCE.getAsyncTimeout(), Config.INSTANCE.getAsyncTimeUnit())
+        );
+
+        assertNotNull(resultDto);
+        assertEquals(EXECUTION_ID, resultDto.getExecutionId());
+        assertSame(ExplainabilityStatus.FAILED, resultDto.getStatus());
+        assertNotNull(resultDto.getStatusDetails());
```
Comment (danielezonca) [inline explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ExplanationServiceImplTest.java]: Please add an assertion on the value too

File: trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java [hunk 3]
```diff
@@ -43,16 +47,29 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+// TODO: improve this test with decision values
```
Comment (danielezonca) [inline trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java]: Wdym?
Btw can you please create a ticket and link it here if you don't want to include these tests in the PR?

File: trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java [hunk 4]
```diff
@@ -75,48 +92,48 @@ void testInvalidPayload() {
...
-        Assertions.assertEquals(resultDto.getExecutionId(), explainabilityResult.getExecutionId());
-        Assertions.assertEquals(resultDto.getSaliencies().size(), explainabilityResult.getSaliencies().size());
-        Assertions.assertTrue(resultDto.getSaliencies().containsKey("saliency"));
-        Assertions.assertEquals(resultDto.getSaliencies().get("saliency").getFeatureImportance().size(),
-                explainabilityResult.getSaliencies().get("saliency").getFeatureImportance().size());
+        Assertions.assertEquals(TEST_RESULT_DTO.getExecutionId(), explainabilityResult.getExecutionId());
+        Assertions.assertEquals(TEST_RESULT_DTO.getSaliencies().size(), explainabilityResult.getSaliencies().size());
+        Assertions.assertTrue(TEST_RESULT_DTO.getSaliencies().containsKey("saliency"));
+        // Assertions.assertEquals(TEST_RESULT_DTO.getSaliencies().get("saliency").getFeatureImportance().size(),
+        // explainabilityResult.getSaliencies().get("saliency").getFeatureImportance().size());
```
Comment (danielezonca) [inline trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java]: Can you please update/re-enable these assertions?

File: trusty/trusty-storage/trusty-storage-api/pom.xml [hunk 5]
```diff
@@ -13,6 +13,10 @@
   <name>Kogito :: Trusty Storage API</name>
 
   <dependencies>
+    <dependency>
+      <groupId>org.kie.kogito</groupId>
+      <artifactId>explainability-api</artifactId>
+    </dependency>
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-api/pom.xml]: From what I can see, the only reason to add this dependency is `ExplainabilityStatus`. Is this really needed? We have `Saliency` and many other classes that are sort of duplicated here so I don't know if it makes sense to add the dependency just for a simple enum. Wdyt?

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java [hunk 6]
```diff
@@ -82,36 +85,43 @@ protected void handleCloudEvent(CloudEventImpl<ExplainabilityResultDto> cloudEve
         LOGGER.info("Received CloudEvent with id {} from {}", attributes.getId(), attributes.getSource());
 
         ExplainabilityResultDto explainabilityResult = optData.get();
+        String executionId = explainabilityResult.getExecutionId();
+        Decision decision = service.getDecisionById(executionId);
```
Comment (danielezonca) [inline trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java]: Can you please add a log message if there are no `Decision` that match with `executionId`?
Can you also add a test for it?

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java (hunk 1 above)
Comment (kostola) [reply explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java]: This is needed imho. The implementation of `LimeExplainer` does not wrap everything in a `CompletableFuture`, but instead does some computation before, that may lead to an exception being thrown (e.g. when the `Prediction` object it receives has no inputs).
With the original implementation, exceptions thrown directly in `explainAsync` wouldn't be catched by the `exceptionally` block and then bubble up to the event loop thread, blocking the whole computation (which means no failure message being sent back to the Trusty Service).
This way, even if the `localExplainer` implementation is not perfect, we are sure to catch every exception it may throw in the `exceptionally` block.

File: explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ExplanationServiceImplTest.java (hunk 2 above)
Comment (kostola) [reply explainability/explainability-service/src/test/java/org/kie/kogito/explainability/ExplanationServiceImplTest.java]: Done :+1:

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java (hunk 1 above)
Comment (kostola) [reply explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java]: As @danielezonca suggested in a chat discussion we had, I now handled this issue using a `try catch` block instead of calling `CompletableFuture::supplyAsync` to avoid overloading the event loop too much.

File: trusty/trusty-storage/trusty-storage-api/pom.xml (hunk 5 above)
Comment (kostola) [reply trusty/trusty-storage/trusty-storage-api/pom.xml]: I removed the dependency and created a dedicated `ExplainabilityStatus` enum in `trusty-storage-api`. When we're done with the first release we may discuss a refactoring of this part.

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java [hunk 7]
```diff
@@ -61,17 +63,17 @@ public ExplanationServiceImpl(
...
-                .thenApply(input -> createResultDto(input, request.getExecutionId()))
-                .exceptionally(throwable -> {
-                    LOG.error("Exception thrown during explainAsync", throwable);
-                    return new ExplainabilityResultDto(request.getExecutionId(), Collections.emptyMap());
-                });
+        try {
+            return localExplainer.explainAsync(getPrediction(request.getInputs(), request.getOutputs()), predictionProvider)
+                    .thenApply(input -> createSucceededResultDto(request.getExecutionId(), input))
+                    .exceptionally(e -> createFailedResultDto(request.getExecutionId(), e));
+        } catch (Exception e) {
```
Comment (r00ta) [inline explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:70]: Is the try catch needed in this case? Is there a test for such scenario?

**Review by r00ta — COMMENTED** [review]
Just a small comment

File: explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java (hunk 7 above)
Comment (kostola) [reply explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:70]: Yes it is needed and there is a test. Please check [this discussion](https://github.com/kiegroup/kogito-apps/pull/428#discussion_r482002768).

Comment (danielezonca) [reply explainability/explainability-service/src/main/java/org/kie/kogito/explainability/ExplanationServiceImpl.java:70]: This catch is to prevent `explainAsync` method to throw an unexpected exception instead of return the `CompletableFuture`

File: trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java (hunk 3 above)
Comment (kostola) [reply trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java]: Sorry that was a leftover. I implemented that part now.

File: trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java (hunk 4 above)
Comment (kostola) [reply trusty/trusty-service/src/test/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumerTest.java]: Done. It needed some refactoring but now it's ok.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java (hunk 6 above)
Comment (kostola) [reply trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java]: Done :+1:

**Review by danielezonca — APPROVED** [review]
LGTM :+1:

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java [hunk 8]
```diff
@@ -82,36 +86,65 @@ protected void handleCloudEvent(CloudEventImpl<ExplainabilityResultDto> cloudEve
...
+
+        service.storeExplainabilityResult(executionId, explainabilityResultFrom(explainabilityResult, decision));
     }
 
-    protected static ExplainabilityResult explainabilityResultFrom(ExplainabilityResultDto dto) {
+    protected Decision getDecisionById(String executionId) {
+        try {
+            return service.getDecisionById(executionId);
+        } catch (IllegalArgumentException e) {
+            return null;
```
Comment (jiripetrlik) [inline trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:103]: Shouldn't we rather throw or exception or return some error info here?

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java [hunk 9]
```diff
@@ -82,36 +86,65 @@ protected void handleCloudEvent(CloudEventImpl<ExplainabilityResultDto> cloudEve
...
-                        .collect(Collectors.toMap(Map.Entry::getKey, e -> saliencyFrom(e.getValue())));
-        return new ExplainabilityResult(dto.getExecutionId(), saliencies);
+                        .map(e -> saliencyFrom(outcomeNameToIdMap.get(e.getKey()), e.getKey(), e.getValue()))
+                        .collect(Collectors.toList());
+        return new ExplainabilityResult(dto.getExecutionId(), statusFrom(dto.getStatus()), dto.getStatusDetails(), saliencies);
     }
 
     protected static FeatureImportance featureImportanceFrom(FeatureImportanceDto dto) {
         if (dto == null) {
             return null;
```
Comment (jiripetrlik) [inline trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:125]: Please consider to throw exception.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java [hunk 10]
```diff
@@ -82,36 +86,65 @@ protected void handleCloudEvent(CloudEventImpl<ExplainabilityResultDto> cloudEve
...
         }
-        return new FeatureImportance(dto.getFeatureId(), dto.getScore());
+        return new FeatureImportance(dto.getFeatureName(), dto.getScore());
     }
 
-    protected static Saliency saliencyFrom(SaliencyDto dto) {
+    protected static Saliency saliencyFrom(String outcomeId, String outcomeName, SaliencyDto dto) {
         if (dto == null) {
             return null;
```
Comment (jiripetrlik) [inline trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:132]: Please consider to throw exception.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java [hunk 11]
```diff
@@ -82,36 +86,65 @@ protected void handleCloudEvent(CloudEventImpl<ExplainabilityResultDto> cloudEve
...
-}
 
+    protected static ExplainabilityStatus statusFrom(org.kie.kogito.explainability.api.ExplainabilityStatus status) {
+        switch (status) {
+            case SUCCEEDED:
+                return ExplainabilityStatus.SUCCEEDED;
+            case FAILED:
+                return ExplainabilityStatus.FAILED;
+        }
+        return null;
```
Comment (jiripetrlik) [inline trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:148]: Please consider to throw exception.

**Review by jiripetrlik — COMMENTED** [review]
I have some comments about exception handling.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java (hunk 8 above)
Comment (kostola) [reply trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:103]: Throwing a `IllegalArgumentException` is the current behavior of the `TrustyServiceImpl` when it doesn't find a decision with the associated ID.
In this case in my implementation I need to return `null` when this happens.
For me it's not an error and in fact it's logged as warning by the caller method because the computation can continue in a "best effort" way even without a `Decision` object.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java (hunk 9 above)
Comment (kostola) [reply trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:125]: It's not an error in my opinion: if the dto for some reason is `null`, then the model should be `null`.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java (hunk 10 above)
Comment (kostola) [reply trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:132]: It's not an error in my opinion: if the dto for some reason is `null`, then the model should be `null`.

File: trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java (hunk 11 above)
Comment (kostola) [reply trusty/trusty-service/src/main/java/org/kie/kogito/trusty/service/messaging/incoming/ExplainabilityResultConsumer.java:148]: It's not an error in my opinion: if the dto for some reason is `null`, then the model should be `null`.

**Comment by danielezonca** [comment]
To be merged together with
https://github.com/kiegroup/kogito-apps/pull/430
(endpoint update)

**Review by jiripetrlik — APPROVED** [review]
Thank you for info. Looks good to me.

**Review by r00ta — APPROVED** [review]
